kong/resty/dns_client/README.md

```diff
@@ -0,0 +1,117 @@
+
+## new
+
+**syntax:** *c, err = dns_client.new(opts)*  
+**context:** any
+
+**Functionality:**
+
+Creates a dns client object. Returns nil and a message string on error.
+
+Perform a series of initialization operations:
+
+* parse `host` file
+* parse `resolv.conf` file (used by the underlying `lua-resty-dns` library)
+* initialize multiple TTL options
+* create a mlcache object and initialize it
+
+**Input paramenters:**
+
+`@opts` It accepts a options table argument. The following options are supported:
+
+* TTL options
+  * `valid_ttl`
+    * same to the option `dns_valid_ttl` in kong.conf
+  * `stale_ttl`
+    * same to the option `dns_stale_ttl` in kong.conf
+  * `empty_ttl`
+    * same to the option `dns_not_found_ttl` in kong.conf
+  * `bad_ttl`
+    * same to the option `dns_error_ttl` in kong.conf
+* `hosts` (default: `/etc/hosts`)
+  * the path of `hosts` file
+* `resolv_conf` (default: `/etc/resolv.conf`)
+  * the path of `resolv.conf` file, it will be parsed and passed into the underlying `lua-resty-dns` library.
+* `order` (default: `{ "LAST", "SRV", "A", "AAAA", "CNAME" }`)
+  * the order in which to resolve different record types, it's similar to the option `dns_order` in kong.conf.
+  * The `LAST` type means the type of the last successful lookup (for the specified name).
+* `enable_ipv6` (default: `ture`)
+  * whether to support IPv6 servers when when getting nameservers from `resolv.conf`
+* options for the underlying `lua-resty-dns` library 
+  * `retrans` (default: `5`)
+    * the total number of times of retransmitting the DNS request when receiving a DNS response times out according to the timeout setting. When trying to retransmit the query, the next nameserver according to the round-robin algorithm will be picked up.
+    * If not given, it is taken from `resolv.conf` option `options attempts:<value>`.
+  * `timeout` (default: `2000`)
+    * the time in milliseconds for waiting for the response for a single attempt of request transmission
+    * If not given, it is taken from `resolv.conf` option `options timeout:<value>`. But note that its unit in `resolv.conf` is second.
+  * `no_random` (default: `true`)
+    * a boolean flag controls whether to randomly pick the nameserver to query first, if `true` will always start with the first nameserver listed. 
+    * If not given, it is taken from `resolv.conf` option `rotate` (inverted).
+  * `nameservers`
+    * a list of nameservers to be used. Each nameserver entry can be either a single hostname string or a table holding both the hostname string and the port number. For exmaple, `{"8.8.8.8", {"8.8.4.4", 53} }`.
+    * If not given, it is taken from `resolv.conf` option `nameserver`.
+
+[Back to TOC](#table-of-contents)
+
+## resolve
+
+**syntax:** *answers, err, tries? = resolve(name, opts?, tries?)*  
```

Comment:
agreed, this is the hot-path. In the actual lookups its a lesser concern since the cache is in between, yet here we're still before the cache-lookups.

Reply:
Oh yea, I'll update it to improve the performance in the hot-path.

Reply:
done, pls recheck this:)